tests/cli/main.cpp

```diff
@@ -0,0 +1,396 @@
+#include <boost/test/included/unit_test.hpp>
+
+/*********************
+ * Helper Methods
+ *********************/
+
+// hack:  import create_example_genesis() even though it's a way, way
+// specific internal detail
+namespace graphene { namespace app { namespace detail {
+	graphene::chain::genesis_state_type create_example_genesis();
+} } } // graphene::app::detail
+
+boost::filesystem::path create_genesis_file(fc::temp_directory& directory) {
+	boost::filesystem::path genesis_path = boost::filesystem::path{directory.path().generic_string()} / "genesis.json";
+    fc::path genesis_out = genesis_path;
+    graphene::chain::genesis_state_type genesis_state = graphene::app::detail::create_example_genesis();
+    fc::json::save_to_file(genesis_state, genesis_out);
+    return genesis_path;
+}
+
+//////
+// @brief attempt to find an available port on localhost
+// @returns an available port number, or -1 on error
+/////
+int get_available_port()
+{
+   struct sockaddr_in sin;
+   int socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+   if (socket_fd == -1)
+      return -1;
+   sin.sin_family = AF_INET;
+   sin.sin_port = 0;
+   sin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+   if (bind(socket_fd, (struct sockaddr*)&sin, sizeof(struct sockaddr_in)) == -1)
+      return -1;
+   socklen_t len = sizeof(sin);
+   if (getsockname(socket_fd, (struct sockaddr *)&sin, &len) == -1)
+      return -1;
+   //return ntohs(sin.sin_port);
+   return sin.sin_port;
+}
+
+/***
+ * @brief Start the application, listening on port 8090
+ * @param app_dir the temporary directory to use
+ * @returns the application object
+ */
+std::shared_ptr<graphene::app::application> start_application(fc::temp_directory& app_dir, int& server_port_number) {
+	std::shared_ptr<graphene::app::application> app1(new graphene::app::application{});
+
+    app1->register_plugin<graphene::account_history::account_history_plugin>();
+    app1->register_plugin< graphene::market_history::market_history_plugin >();
+    app1->register_plugin< graphene::witness_plugin::witness_plugin >();
+    app1->startup_plugins();
+    boost::program_options::variables_map cfg;
+    server_port_number = get_available_port();
+    //server_port_number = 8090;
+    cfg.emplace("rpc-endpoint", boost::program_options::variable_value(string("127.0.0.1:" + std::to_string(server_port_number)), false));
+    cfg.emplace("genesis-json", boost::program_options::variable_value(create_genesis_file(app_dir), false));
+    cfg.emplace("seed-nodes", boost::program_options::variable_value(string("[]"), false));
+    app1->initialize(app_dir.path(), cfg);
+
+    app1->startup();
+    fc::usleep(fc::milliseconds(500));
+	return app1;
+}
+
+/****
+ * Send a block to the db
+ * @param app the application
+ * @returns true on success
+ */
+bool generate_block(std::shared_ptr<graphene::app::application> app) {
+	try {
+	    fc::ecc::private_key committee_key = fc::ecc::private_key::regenerate(fc::sha256::hash(string("nathan")));
```

Comment:
Can use the same method to generate a new key and replace the keys generated in genesis file.

Reply:
Still working on this...